src/encrypt.c

```diff
@@ -220,22 +226,33 @@ ece_webpush_encrypt_plaintext(
     padLen -= blockPadLen;
 
     // Fill the rest of the block with plaintext.
-    size_t plaintextEnd = plaintextStart + dataPerBlock - blockPadLen;
-    bool plaintextExhausted = false;
-    if (plaintextEnd >= plaintextLen) {
+    size_t maxBlockPlaintextLen = dataPerBlock - blockPadLen;
+    size_t plaintextEnd;
+    if (maxBlockPlaintextLen > plaintextLen - plaintextStart) {
+      // Equivalent to `plaintextStart + maxBlockPlaintextLen > plaintextLen`
+      // without overflow.
       plaintextEnd = plaintextLen;
-      plaintextExhausted = true;
+    } else {
+      plaintextEnd = plaintextStart + maxBlockPlaintextLen;
     }
     size_t blockPlaintextLen = plaintextEnd - plaintextStart;
 
     // The full length of the plaintext and padding.
     size_t blockLen = blockPlaintextLen + blockPadLen;
 
-    size_t ciphertextEnd = ciphertextStart + blockLen + overhead;
-    if (ciphertextEnd > maxCiphertextLen) {
+    // The full length of the encrypted record, including room for the
+    // plaintext, padding, padding delimiter, and auth tag.
+    size_t recordLen = blockLen + overhead;
```

Comment:
I renamed `dataPerBlock` to `maxBlockLen`, to make it clearer that `blockLen` should be <= `maxBlockLen`, and added some comments and assertions. Here's my reasoning:

* `maxBlockLen = rs - overhead`. If `rs = UINT32_MAX`, and `sizeof(size_t) == sizeof(uint32_t)`, `maxBlockLen = SIZE_MAX - (17 || 18)`. Also, we check and fail if `rs <= overhead`, so this shouldn't underflow.
* `blockPadLen = min(max(maxBlockLen - 1, 1), padLen)`, so we know `blockPadLen <= padLen`, and `blockPadLen <= maxBlockLen`. `blockPadLen == maxBlockLen` if `rs = overhead + 1`; otherwise, `blockPadLen < maxBlockLen`.
* `blockPlaintextLen = plaintextEnd - plaintextStart`. `blockPlaintextLen = 0` if `rs = overhead + 1`, or if we're writing a trailer.
* `maxBlockPlaintextLen = maxBlockLen - blockPadLen`.
* `blockPlaintextLen <= maxBlockPlaintextLen`, so we know `blockPlaintextLen + blockPadLen` is safe. (It should add up to `maxBlockLen`, except for the last record, or if we have too much padding).
* `blockLen <= maxBlockLen`, so we know `blockLen + overhead` is safe. (It should add up to `rs`, except for the last record).